### ISSUE TITLE
Provide native support for using DKS (Domain Keystore) files in application.properties

### ISSUE THREAD

**Issue by LawOfSynergy** [spring-projects/spring-boot#13590]
# What I would Expect

Given an application.properties file that looks like the following:

```
server.port=8080
server.ssl.enabled=true
server.ssl.key-store=certs/dev-dks.dks
server.ssl.keyStoreType: DKS
```

I would expect the application context to be properly initialized using a `DomainLoadStoreParameter` object.

# What currently happens 

```
java.lang.UnsupportedOperationException: This keystore must be loaded using a DomainLoadStoreParameter
```

I am not sure if this is considered a bug, but it would definitely be a useful feature to configure an application to load a dks without having to perform any ssl configuration in the codebase.

Thank you for your time and consideration!

**Comment by philwebb** [comment]
The keystore type is usually passed down to the underling server to handle. For example, with Tomcat we call `AbstractHttp11Protocol.setKeystoreType(...)`. What's the full stacktrace for the exception that you're seeing?

**Comment by LawOfSynergy** [comment]
```
java.lang.UnsupportedOperationException: This keystore must be loaded using a DomainLoadStoreParameter
	at sun.security.provider.DomainKeyStore.engineLoad(DomainKeyStore.java:713) ~[na:1.8.0_111]
	at sun.security.provider.DomainKeyStore$DKS.engineLoad(DomainKeyStore.java:68) ~[na:1.8.0_111]
	at java.security.KeyStore.load(KeyStore.java:1445) ~[na:1.8.0_111]
	at org.apache.tomcat.util.net.SSLUtilBase.getStore(SSLUtilBase.java:136) ~[tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.tomcat.util.net.SSLHostConfigCertificate.getCertificateKeystore(SSLHostConfigCertificate.java:187) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.tomcat.util.net.jsse.JSSEUtil.getKeyManagers(JSSEUtil.java:185) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.tomcat.util.net.AbstractJsseEndpoint.createSSLContext(AbstractJsseEndpoint.java:112) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.tomcat.util.net.AbstractJsseEndpoint.initialiseSsl(AbstractJsseEndpoint.java:85) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.tomcat.util.net.NioEndpoint.bind(NioEndpoint.java:225) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.tomcat.util.net.AbstractEndpoint.start(AbstractEndpoint.java:978) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.coyote.AbstractProtocol.start(AbstractProtocol.java:628) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.catalina.connector.Connector.startInternal(Connector.java:993) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.apache.catalina.core.StandardService.addConnector(StandardService.java:225) [tomcat-embed-core-8.5.14.jar!/:8.5.14]
	at org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer.addPreviouslyRemovedConnectors(TomcatEmbeddedServletContainer.java:247) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer.start(TomcatEmbeddedServletContainer.java:190) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.startEmbeddedServletContainer(EmbeddedWebApplicationContext.java:297) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.finishRefresh(EmbeddedWebApplicationContext.java:145) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:545) [spring-context-4.3.8.RELEASE.jar!/:4.3.8.RELEASE]
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:737) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:370) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:314) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1162) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1151) [spring-boot-1.5.3.RELEASE.jar!/:1.5.3.RELEASE]
	at [Redacted]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_111]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_111]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_111]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_111]
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) [giant2-ccp-example.jar:1.19.20-SNAPSHOT]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) [giant2-ccp-example.jar:1.19.20-SNAPSHOT]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50) [giant2-ccp-example.jar:1.19.20-SNAPSHOT]
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51) [giant2-ccp-example.jar:1.19.20-SNAPSHOT]
```

**Comment by wilkinsona** [comment]
Thanks for the stacktrace.

I don't think it's possible to use DKS with Tomcat as it does not provide sufficient control over how the KeyStore is loaded. It needs to read a key store from a file or by opening a URL connection. You could try using an `SslStoreProvider` bean, but I suspect that will not work either as we have to write the KeyStore out to an in-memory byte array so that Tomcat can read it back in from a connection to a URL with a custom protocol. If you'd like it to work with Tomcat, I'd recommend opening a Tomcat enhancement request [here](https://bz.apache.org/bugzilla/) to see if Tomcat can add an API for directly configuring the `KeyStore`.

It should work with an `SslStoreProvider`if you use Jetty or Undertow as we can just pass the `KeyStore` from the provider straight to the container.

I don't think we should try to provide support for using DKS via `application.properties`. We'd have to provide additional configuration options for creating the `DomainLoadStoreParameter` and it's not clear how we could do that for the map of protection parameters.

**Comment by LawOfSynergy** [comment]
A note for the future:

If you were to move forward with providing support for DKS using application.properties, I would expect a reasonable assumption is that the passwords are provided through the environment using `keystorePasswordEnv` entries in the dks file itself. Is there a reason there may be a more complex use case where they need to set those env variables at runtime but before the configuration step is completed? I would expect running an initialization script before starting the application should be sufficient for the majority of purposes. In fact, I would expect it to be best practice for a script to securely pull such passwords from a password vault (such as hashicorp) and store them as environment variables for production clusters. Perhaps that is unwieldy for smaller/non-managed projects? I am very much interested if you guys have any feedback, comments, or speculations on this.

In the meantime, I will go ahead and request the tomcat enhancement.

**Comment by wilkinsona** [comment]
You make some good points. Thanks.

> Perhaps that is unwieldy for smaller/non-managed projects?

My gut feel is that smaller/non-managed projects won't be interested in DKS. Perhaps we could limit ourselves to creating `DomainLoadStoreParameter` with an empty map. That would allow use to support DKS from application.properties with Jetty and Undertow. I think we'd still be stuck for Tomcat though as we can't stop it from calling `java.security.KeyStore.load(InputStream, char[])`.

**Comment by LawOfSynergy** [comment]
In that case, if tomcat does end up implementing a change to their api to allow for more control over keystore loading, should I leave a comment here, and you will reopen the ticket? Or should I create a new ticket to represent the feature request?

**Comment by wilkinsona** [comment]
On reflection, I've re-opened this issue. The necessary change seems to be straightforward with Undertow:

```diff
diff --git a/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/web/embedded/undertow/SslBuilderCustomizer.java b/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/web/embedded/undertow/SslBuilderCustomizer.java
index eb0cbdbdad..ee40333a46 100644
--- a/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/web/embedded/undertow/SslBuilderCustomizer.java
+++ b/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/web/embedded/undertow/SslBuilderCustomizer.java
@@ -19,12 +19,14 @@ package org.springframework.boot.web.embedded.undertow;
 import java.net.InetAddress;
 import java.net.Socket;
 import java.net.URL;
+import java.security.DomainLoadStoreParameter;
 import java.security.KeyManagementException;
 import java.security.KeyStore;
 import java.security.NoSuchAlgorithmException;
 import java.security.Principal;
 import java.security.PrivateKey;
 import java.security.cert.X509Certificate;
+import java.util.Collections;

 import javax.net.ssl.KeyManager;
 import javax.net.ssl.KeyManagerFactory;
@@ -181,7 +183,13 @@ class SslBuilderCustomizer implements UndertowBuilderCustomizer {
                }
                KeyStore store = KeyStore.getInstance(type);
                URL url = ResourceUtils.getURL(resource);
-               store.load(url.openStream(), password != null ? password.toCharArray() : null);
+               if ("DKS".equalsIgnoreCase(type)) {
+                       store.load(new DomainLoadStoreParameter(url.toURI(), Collections.emptyMap()));
+               }
+               else {
+                       store.load(url.openStream(),
+                                       password != null ? password.toCharArray() : null);
+               }
                return store;
        }
```

I imagine Jetty and Netty will be similarly straightforward. We can consider implementing this now for Jetty, Netty, and Undertow and documenting the limitation with Tomcat for the time being.

When the Tomcat enhancement is available please:

- Comment on this issue if it's still open or has been closed without a change being made
- Open a new issue if this issue has been used to implement an enhancement for the three other containers

**Comment by LawOfSynergy** [comment]
Thank you! I will do that!

**Comment by LawOfSynergy** [comment]
FYI, I have submitted the feature request: https://bz.apache.org/bugzilla/show_bug.cgi?id=62507

**Comment by LawOfSynergy** [comment]
The feature request has been fixed. I do not know if any action is required on Spring Boot's part considering the change they made.

**Comment by wilkinsona** [comment]
We don't need to do anything for Tomcat (other than to upgrade to the new version when it's released). For Jetty, Netty, and Undertow we'll need to make changes similar to the one I [outlined above](https://github.com/spring-projects/spring-boot/issues/13590#issuecomment-401094392).

**Comment by wilkinsona** [comment]
Jetty isn't as straightforward as I'd hoped. When using Jetty we configure SSL using an instante of `org.eclipse.jetty.util.ssl.SslContextFactory`. As far as I can tell, it doesn't supported DKS stores. However, you can still use DKS with Jetty by providing an `SslStoreProvider` and creating the `KeyStore` yourself. The same approach will also work with Netty and Undertow.

This leaves Tomcat as an odd one out as you can configure a DKS store with `application.properties` by setting the store type. We could also support this with Undertow and Netty with minimal changes, but supporting it with Jetty would require a rewrite of how SSL is configured so that we create the key and trust stores and set them on `SslContextFactory`, rather than configuring it and having it create them.

Given that we haven't seen much demand for DKS support beyond this request and that it's now possible to use DKS across all four containers one way or another, I'm going to close this one.

In summary, to use DKS with Tomcat you should set the store type to DKS in application.properties and provide a path point to the store. To use DKS with Jetty, Netty, and Undertow, an `SslStoreProvider` should be provided that returns the DKS `KeyStore`.
